### PR TITLE
Add EIP-3712: Standard for Multiple Types of Fungible-Tokens

### PR DESCRIPTION
**Pull request by 1999321**
When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

You can find it from :
- https://github.com/naturaldao/EIPs/blob/Eip-Branch/EIPS/eip-3712.md

### CODE REVIEW

File: EIPS/eip-3712.md
Comment (lightclient) [inline EIPS/eip-3712.md:2]:
```suggestion
eip: 3712
```

File: EIPS/eip-3712.md
Comment (lightclient) [inline EIPS/eip-3712.md:3]:
```suggestion
title: Multiple Fungible Token Standard
```

File: EIPS/eip-3712.md
Comment (lightclient) [inline EIPS/eip-3712.md:5]: Will need to add a discussions link (usually on ethereum-magicians.org) before merge.

File: EIPS/eip-3712.md
Comment (lightclient) [inline EIPS/eip-3712.md:18]: Please link to ERCs you're referencing as specified in EIP-1 like this: `[ERC-20](./eip-20.md)`.

File: EIPS/eip-3712.md
Comment (lightclient) [inline EIPS/eip-3712.md:26]:
```suggestion
Interfaces: eip.sol
```
?

File: EIPS/eip-3712.md
Comment (lightclient) [inline EIPS/eip-3712.md:318]:
```suggestion
## Copyright
```

File: EIPS/eip-3712.md
Comment (lightclient) [inline EIPS/eip-3712.md:87]: Please specify if you EIP is backwards compatible with other ERC standards and the Ethereum protocol.

**Review by lightclient — CHANGES_REQUESTED** [review]

File: EIPS/eip-3712.md
Comment (wschwab) [reply EIPS/eip-3712.md:18]: You may also want to consider reframing your goals. The functions provided in this standard are essentially an extension to 1155 - you are keeping all of 1155's functions, and then adding new functionality on top of it afaict. You could frame this as an extension of 1155, `require` it in the beginning, and significantly simplify the specification and implementation. Would that be in line with your goals for this EIP?

File: EIPS/eip-3712.md
Comment (wschwab) [inline EIPS/eip-3712.md:22]: It might be worth noting that 1155 technically already supports fungible tokens. This may be an incorrect assumption on my part, but I would assume the reason that 1155 never really took off for ERC20-like tokens is due to the difficulty of integrating them on the smart contract/app layer - web3 code assumes that ERC20-like tokens are one to a contract (and therefore do not need a `tokenId` when being referenced), as do Solidity functions integrating with them. For example, Uniswap (to the best of my knowledge) does not have any ability to integrate a token with a `tokenId` into the pool.

As a result, imho it would be hard to see this EIP getting wide traction in the form you intend (not unlike 1155). I don't mean this to be discouraging - I think the functions that you propose are very useful, and in fact know of contracts implementing similar functions for convenience. imho you might want to change the motivation to reflect that.

File: EIPS/eip-3712.md
Comment (1999321) [reply EIPS/eip-3712.md]: > Please specify if you EIP is backwards compatible with other ERC standards and the Ethereum protocol.

Thank you for your suggestion. I just submitted the changes.

File: EIPS/eip-3712.md
Comment (zhous) [reply EIPS/eip-3712.md]: I'm a team member of this EIP. The application goal (www.DAism.io) of this EIP is to build a platform for Ethereum economy, by creating a band new way to issue tokens and a band new decentralized exchange network (we call it IADD), which is obviously a replacement of Uniswap. 
On Nov 13th, 2019,  I wrote an article relating to NFT (https://deme.app/full#12), But I emilinated it from the core of DAism the next year. So we only focus on fungile token now.
It's a very good suggestion to modify the motivation. Thank you very much!

Comment (1999321) [reply EIPS/eip-3712.md]: Thank you all for your inputs. Yes what "wschwab" said pretty makes sense. However when utilizing this standard we can handling multiple fungible tokens in a single contract, and it makes transactions of multiple fungible tokens much easier and more gas-efficient. In addition ERC-1155 is more widely used in games while this standard can be more widely used in DeFi. This standard partly inherits some functions in transaction from ERC-1155 but has new functions specifically to handle multipe fungile tokens. ERC-1155 doesn't differentiate fungible tokens from non fungible tokens, doesn't have attributes such as precisions, token name, token symbol to define a fungible token, and doesn't have functions to handle different kinds of approvals for fungible tokens (it only has a function to set a global approval). ERC-1155's overall design is more intended for games. It doesn't cover many use cases such as DeFi in which multiple fungible tokens need to be handled more efficiently. Therefore we come up with our EIP to cover these use cases. We inherit the functions and features of ERC-1155, which apply to fungible tokens and add more attributes such as ID and functions such as different approval functions to better handle multiple fungible tokens in more general use cases. Our proposed EIP is intended for handling multiple fungible tokens in wider scenarios and in a more efficient way.

File: EIPS/eip-3712.md
Comment (lightclient) [inline EIPS/eip-3712.md:12]: We recently deprecated the "Simple Summary" in favor of a description field in the preamble. Please see the [EIP template ](https://github.com/ethereum/EIPs/blob/master/eip-template.md)for an example.

**Comment by Reki5** [comment]
> When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md
> 
> We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:
> 
> * The PR edits only existing draft PRs.
> * The build passes.
> * Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside .
> * If matching on email address, the email address is the one publicly listed on your GitHub profile.
> 
> You can find it from :
> 
> * https://github.com/naturaldao/EIPs/blob/Eip-Branch/EIPS/eip-3712.md

To Reki5

File: EIPS/eip-3712.md
Comment (lightclient) [inline EIPS/eip-3712.md:4]:
```suggestion
author: Zhao Li (@1999321), Derek Zhou (@zhous), Yuefei Tan (@whtyfhas)  
```

**Review by lightclient — CHANGES_REQUESTED** [review]

File: EIPS/eip-3712.md
Comment (1999321) [reply EIPS/eip-3712.md]: Thank you all for your inputs.

File: EIPS/eip-3712.md
Comment (1999321) [reply EIPS/eip-3712.md]: Thank you all for your inputs

File: EIPS/eip-3712.md
Comment (lightclient) [inline EIPS/eip-3712.md:4]:
```suggestion
author: Zhao Li (@1999321), Derek Zhou (@zhous), Yuefei Tan (@whtyfhas)
```

Comment (lightclient) [inline EIPS/eip-3712.md:4]:
```suggestion
author: Zhao Li (@1999321), Derek Zhou (@zhous), Yuefei Tan (@whtyfhas)
```

**Review by lightclient — CHANGES_REQUESTED** [review]

File: EIPS/eip-3712.md
Comment (SamWilsn) [inline EIPS/eip-3712.md:3]: There's a (relatively) new preamble section required: [`description`](https://eips.ethereum.org/EIPS/eip-1#eip-header-preamble).

```suggestion
title: Multiple Fungible Token Standard
description: standardize single contracts that support multiple fungible tokens
```

File: EIPS/eip-3712.md
Comment (SamWilsn) [inline EIPS/eip-3712.md:74]:
```suggestion
`approve(address spender,bool status)` is similar to [ERC-1155](./eip-1155.md)'s interface. It allows an approval to be set uniformly for all types of tokens.
```

File: EIPS/eip-3712.md
Comment (SamWilsn) [inline EIPS/eip-3712.md:76]: Not a strict requirement, but I think it would be better to define how these functions work without reference to EIP-1155.

**Review by SamWilsn — COMMENTED** [review]
Sorry it's taken so long to come back and review this!

File: EIPS/eip-3712.md
Comment (1999321) [reply EIPS/eip-3712.md:77]: thanks for your advice.

File: EIPS/eip-3712.md
Comment (SamWilsn) [inline EIPS/eip-3712.md:333]: The only blocker that I see left to get this merged is that the `Security Considerations` section is missing and is not optional.

**Review by SamWilsn — CHANGES_REQUESTED** [review]
Still needs a `Security Considerations` section.

File: EIPS/eip-3712.md
Comment (Pandapip1) [inline EIPS/eip-3712.md:336]:
```suggestion
Copyright and related rights waived via [CC0](../LICENSE.md).
```

File: EIPS/eip-3712.md
Comment (Pandapip1) [inline EIPS/eip-3712.md:10]:
```suggestion
created: 2021-08-01
requires: 20, 721, 1155
```

File: EIPS/eip-3712.md
Comment (Pandapip1) [inline EIPS/eip-3712.md:28]:
```suggestion
interface IEIP3712 {
```

File: EIPS/eip-3712.md
Comment (Pandapip1) [inline EIPS/eip-3712.md:81]: These should ideally be done in NatSpec comments in the interface.

File: EIPS/eip-3712.md
Comment (Pandapip1) [inline EIPS/eip-3712.md:86]:
```suggestion
The [ERC-20](./eip-20.md) token standard defines a single fungible token. The [ERC-1155](./eip-1155.md) token standard defines multiple types of tokens and specifies each token type by using a parameter `id`. By using this parameter `id`, the [ERC-1155](./eip-1155.md) token standard greatly reduces the gas consumption in transactions utilizing multiple distinct types of tokens, when compared with the [ERC-20](./eip-20.md) token standard. Our proposed standard likewise uses an `id` parameter.  In our proposed standard, since each type of fungible token is specified by an id, we add a function `approve(uint256 id, address spender, uint256 amount)` for applications to set an approval for the specified token type. In order to set approvals for all types of fungible tokens simultaneously, we introduce a function `approve(address spender, bool _status)`  to set approvals uniformly for all types of fungible tokens in a single call. To handle a transaction of multiple types of fungible tokens from one sending address to multiple receiving addresses we introduce an interface `transferBatchMul`. Finally, to handle a transaction of multiple types of fungible tokens from multiple sending addresses to multiple receiving addresses we introduce an interface `transferFromBatchMul`.     
```

File: EIPS/eip-3712.md
Comment (Pandapip1) [inline EIPS/eip-3712.md:90]:
```suggestion
This EIP is partially backwards compatible with [EIP-1155](./eip-1155.md), but is NOT compatible with [EIP-721](./eip-721.md) or [EIP-20](./eip-20.md).
```

I suggest adding that it is not compatible with EIP-721 because it is also a commonly-used token standard (despite being a subset of EIP-1155).

File: EIPS/eip-3712.md
Comment (Pandapip1) [inline EIPS/eip-3712.md:99]:
```suggestion
```

Not sure where this came from

**Review by Pandapip1 — CHANGES_REQUESTED** [review]

File: EIPS/eip-3712.md
Comment (SamWilsn) [inline EIPS/eip-3712.md:4]: We try to discourage standard (and similar words) in the title and description fields. They tend to show up in every EIP otherwise.

**Comment by Pandapip1** [comment]
Closing since the author hasn't been seen in >3 months (the stale threshold).

**Comment by zhous** [comment]
This is still alive but no more modifications at present.

**Pull request closed without merging**
